ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/MnecViolationCostEvaluator.java

```diff
@@ -0,0 +1,92 @@
+import static com.farao_community.farao.commons.Unit.MEGAWATT;
+
+/**
+ * An evaluator that computes the virtual cost resulting from the violation of
+ * the MNEC minimum margin soft constraint
+ *
+ * @author Peter Mitri {@literal <peter.mitri at rte-france.com>}
+ */
+public class MnecViolationCostEvaluator implements CostEvaluator {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MnecViolationCostEvaluator.class);
+
+    private Unit unit;
+    private double mnecAcceptableMarginDiminution;
+    private double mnecViolationCost;
+
+    public MnecViolationCostEvaluator(Unit unit, double mnecAcceptableMarginDiminution, double mnecViolationCost) {
+        if ((unit != MEGAWATT) && (unit != AMPERE)) {
+            throw new NotImplementedException("MNEC violation cost is only implemented in MW and AMPERE units");
+        }
+        this.unit = unit;
+        this.mnecAcceptableMarginDiminution = mnecAcceptableMarginDiminution;
+        this.mnecViolationCost = mnecViolationCost;
+    }
+
+    @Override
+    public double getCost(RaoData raoData) {
+        if (Math.abs(mnecViolationCost) < 1e-10) {
+            return 0;
+        }
+        double totalMnecMarginViolation = 0;
+        boolean mnecsSkipped = false;
+        String initialVariantId =  raoData.getCrac().getExtension(ResultVariantManager.class).getPreOptimVariantId();
+        for (Cnec cnec : raoData.getCrac().getCnecs()) {
+            if (cnec.isMonitored()) {
+                double initialFlow = (unit == MEGAWATT) ? cnec.getExtension(CnecResultExtension.class).getVariant(initialVariantId).getFlowInMW()
+                        : cnec.getExtension(CnecResultExtension.class).getVariant(initialVariantId).getFlowInA();
+                if (Double.isNaN(initialFlow)) {
+                    // Sensitivity results are not available, skip cnec
+                    // (happens on search tree rao rootleaf evaluation)
+                    mnecsSkipped = true;
+                    continue;
+                }
+                double initialMargin = cnec.computeMargin(initialFlow, unit);
+                double newFlow = (unit == MEGAWATT) ? raoData.getSystematicSensitivityAnalysisResult().getReferenceFlow(cnec) :
+                        raoData.getSystematicSensitivityAnalysisResult().getReferenceIntensity(cnec);
+                double newMargin = cnec.computeMargin(newFlow, unit);
+                double convertedAcceptableMarginDiminution = mnecAcceptableMarginDiminution / getUnitConversionCoefficient(cnec, raoData);
+                totalMnecMarginViolation += Math.max(0, Math.min(0, initialMargin - convertedAcceptableMarginDiminution) - newMargin);
+            }
+        }
+        if (mnecsSkipped) {
+            LOGGER.warn("Some MNECs were skipped during violation cost evaluation because their sensitivity results were not available.");
+        }
+        return mnecViolationCost * totalMnecMarginViolation;
```

Comment:
I'm not sure this is consistent with what is in the LP.

If I'm not wrong, here the violation cost is proportional to the margin in A or MW (depending on the unit in the objective function). In the LP it is proportional to the margin in MW.

Maybe it would be easier to say : the mnecViolationCost is always in {unit of the objective function} / MW. That is to say we always multiply it with MW value.
It might also avoid some problem later when the objective function will be neither MW neither A (for instance with the relative margin)

Reply:
mnecViolationCost has no real unit : it is either in "equivalent A virtual cost per A constraint violation", or "equivalent MW virtual cost per MW constraint violation"
so i think it is possible to use its value both in the evaluator and in the LP without worrying about units

Reply:
I agree but we would be less coherent with the functional part of the objective if we change this (MinMarginEvaluator)

Reply:
will change the objective function in LP to use AMPERES when needed

Reply:
perfect